src/main/scala/org/constellation/lb/Manager.scala

```diff
@@ -231,3 +236,39 @@ class Manager(init: NonEmptyList[Addr], config: LoadbalancerConfig)(
       )
   }
 }
+
+object Manager {
+
+  def findNewHosts(clusterInfo: NonEmptyMap[Addr, Option[List[Info]]]): SortedSet[Addr] =
+    SortedSet[Addr]() ++ clusterInfo.toNel.foldLeft(List.empty[Addr])(
+      (acc, bcc) =>
+        bcc match {
+          case (_, Some(hosts: List[Info])) => acc ++ hosts.map(n => Addr(n.ip, n.publicPort)).filterNot(clusterInfo(_).isDefined)
+          case _                            => acc
+        }
+    )
+
+  /**
+   * Split a map of address -> nodes, into two maps: one with the addresses with nodes in the given cluster session,
+   * and other with the addresses with nodes not in the given cluster session.
+   * If knownSession is none, the input map is consider all valid, and returned in the first element of the tuple
+   *
+   * @param knownSession cluster session used to split
+   * @param currentHosts address -> nodes to split
+   * @return (map with nodes in the given session, map with nodes not in the given session)
+   */
+  def selectSameSessionNodes(knownSession: Option[Long], currentHosts: NonEmptyMap[Addr, Option[List[Info]]]): (SortedMap[Addr, Option[List[Info]]], SortedMap[Addr, Option[List[Info]]]) =
+    currentHosts.toSortedMap.partition { case (_, v) => knownSession.flatMap(session => v.map(_.forall(_.clusterSession === session))).getOrElse(true) }
+
+  /**
+   * Finds the first cluster session from a map of address -> nodes
+   *
+   * @param hosts address to nodes map
+   * @return cluster session id found if any
+   */
+  def findSession(hosts: NonEmptyMap[Addr, Option[List[Info]]]): Option[Long] =
+    hosts.find(_.exists(_.nonEmpty)).flatMap(_.flatMap(_.headOption).map(_.clusterSession))
```

Comment:
Can be simplified in a single `collectFirst` call. There is no need to use `headOption` since you already filtered the `empty` ones:

```scala
hosts.toSortedMap.collectFirst {  case (_, Some(info :: _) => info.clusterSession }
```